### PR TITLE
Model and server

### PR DESCRIPTION
**Pull request by hiqua**
N.B.: this includes the commits from #2 so that the GH actions can actually be used. It'd be better to keep the reviews separate though (and I don't mind rebasing).

Notably does not include:
* client TLS (and proper client identification)
* client

Need to address #3 in this PR.

### CODE REVIEW

**Comment by hiqua** [comment]
@r0mant @russjones ready for review.

File: internal/job_directory.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+)
+
+// JobDirectory provides a safe interface to the map containing all jobs for all users.
+type JobDirectory struct {
+	jobs map[model.UserID]map[model.JID]*cmd.Cmd
+	// One global lock. Other designs could perform.
+	lock sync.RWMutex
+}
+
+var jobDirectory = JobDirectory{jobs: make(map[model.UserID]map[model.JID]*cmd.Cmd), lock: sync.RWMutex{}}
```
Comment (russjones) [inline internal/job_directory.go:19]: This probably shouldn't be a global variable. For example, you can do something like the following.

```go
type directory struct {
  mu sync.RWMutex
}

func NewDirectory() (*directory, error) {
  return &directory{
     ...
  }, nil
}

func (d *directory) AddJob(...) string {
 ...
}

...
```

Then call `NewDirectory` early in process startup and pass it along.

File: internal/job_directory.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+	userJobs, ok := jobDirectory.jobs[uid]
+	if !ok {
+		userJobs = make(map[model.JID]*cmd.Cmd)
+		jobDirectory.jobs[uid] = userJobs
+	}
+
+	_, jidExists := userJobs[jid]
+	if jidExists {
+		panic("We assume that the JID are uniquely generated, and that cmd are write-once.")
```
Comment (russjones) [inline internal/job_directory.go:119]: Don't panic and crash in the library, instead return an error and let the caller decide what to do.

File: internal/server.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+	router.POST("/job", addJob)
+	router.DELETE("/stop", stopJob)
+
+	// For debugging purposes.
+	router.GET("/hello/:name", Hello)
+
+	// TODO: better location for these files?
+	// TODO: listen to 0.0.0.0 once mTLS setup
+	log.Fatal(listenAndServeTLS("127.0.0.1:8443", "certs/server/cert.pem", "certs/server/key.pem", router))
```
Comment (russjones) [inline internal/server.go:37]: Don't panic and crash in the library, instead return an error and let the caller decide what to do.

File: internal/server.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// Hello says hello to TLS connection to help debugging.
+func Hello(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	fmt.Fprintf(w, "hello, %s!\n", ps.ByName("name"))
+}
```
Comment (russjones) [inline internal/server.go:43]: Probably not needed anymore right?

File: internal/server.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+// Taken from the http package
+// TODO: maybe there's a better way to enforce TLS1.3 rather than rewrite this function?
+func listenAndServeTLS(addr, certFile, keyFile string, handler http.Handler) error {
+	server := &http.Server{Addr: addr, Handler: handler}
+	config := &tls.Config{}
+	config.MinVersion = tls.VersionTLS13
```
Comment (russjones) [inline internal/server.go:50]:
```go
config := &tls.Config{
  MinVersion: tls.VersionTLS13,
}
```

File: internal/server_utils.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+func sendJSON(w http.ResponseWriter, jsonObject interface{}) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	bs, err := json.MarshalIndent(jsonObject, "", "  ")
+	if err != nil {
+		bs, err = json.Marshal(map[string]string{"err": err.Error()})
+		if err != nil {
+			panic("unreachable")
+		}
+		http.Error(w, string(bs), http.StatusInternalServerError)
```
Comment (russjones) [inline internal/server_utils.go:27]: Won't everything be a 500 here? I think the API definition you provided was potentially returning different error codes.

File: internal/server_utils.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	sendJSON(w, map[string]interface{}{"err": err.Error()})
+}
+
+func sendJSON(w http.ResponseWriter, jsonObject interface{}) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	bs, err := json.MarshalIndent(jsonObject, "", "  ")
+	if err != nil {
+		bs, err = json.Marshal(map[string]string{"err": err.Error()})
+		if err != nil {
+			panic("unreachable")
```
Comment (russjones) [inline internal/server_utils.go:25]: Don't panic.

File: internal/server_utils.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+		}
+		http.Error(w, string(bs), http.StatusInternalServerError)
+		return
+	}
+	fmt.Fprintf(w, "%s\n", bs)
```
Comment (russjones) [inline internal/server_utils.go:30]: Is the new line needed?

File: pkg/model.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+// JobStatus reports on all relevant information about the job (but the logs).
+type JobStatus struct {
+	ID        JID      `json:"id"`
+	Command   string   `json:"command"`
+	Arguments []string `json:"arguments"`
+	Status    Status   `json:"status"`
+	// Can be -1 in the stopped case.
+	ExitCode *int `json:"exitCode,omitempty"`
```
Comment (russjones) [inline pkg/model.go:36]: Why is this a pointer?

Comment (hiqua) [reply pkg/model.go:36]: So that the exitCode value can be omitted when not relevant. In this case, the pointer will be nil, and the `omitempty` will make sure the field is not present.

This advice can be found e.g. in: https://www.sohamkamani.com/golang/omitempty/ 

Let me know if you'd prefer another design.

File: internal/server_utils.go (hunk 7 above)
Comment (hiqua) [reply internal/server_utils.go:25]: This really is unreachable code, since Marshal cannot fail with a `string` (which `err.Error()` is). Basically I meant this code as an `assert(err == nil)` in other languages (if or as the `unreachable` macro in Rust (https://doc.rust-lang.org/std/macro.unreachable.html).

The other alternative I see is `bs, _ = ...` and ignore the returned error, I just found this `panic("unreachable")` more explicit.

So I see two options:
* add a comment
* ignore the error as described

Maybe you know a 3rd better option you'd like to share?

File: internal/job_directory.go (hunk 2 above)
Comment (hiqua) [reply internal/job_directory.go:119]: The JID is generated within the library without asking the caller, and is (should be) unique by design (see `generateJID`). These 3 lines are meant as an `assert`. Should I simply remove them? Or replace them with a comment like "we guarantee that we're not forgetting a job here since we're using a new JID everytime"?

File: internal/server_utils.go (hunk 6 above)
Comment (hiqua) [reply internal/server_utils.go:41]: Good point, I forgot about this. Should be fixed now.

File: internal/job_directory.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+	command, err := d.retrieveJob(uid, jid)
+	if err != nil {
+		return nil, err
+	}
+
+	status := command.Status()
+	stdout, stderr := strings.Join(status.Stdout, "\n"), strings.Join(status.Stderr, "\n")
+
+	// The output is not exact, we miss possibly one \n at the end (at least).
+	return &model.JobLog{Stdout: stdout, Stderr: stderr}, nil
```
Comment (russjones) [inline internal/job_directory.go:74]: Can you clarify the "at least" part?

File: internal/job_directory.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+	// no lock needed, .Status() safe to call concurrently
+	status := command.Status()
+
+	jobStatus := model.JobStatus{ID: jid, Command: command.Name, Arguments: command.Args}
+
+	// See doc of the Status struct: we consider the job done if it has run and
+	// then stopped.
+	if status.StartTs > 0 && status.StopTs > 0 {
+		if status.Exit == -1 {
```
Comment (russjones) [inline internal/job_directory.go:47]: Is this documented behavior for `github.com/go-cmd/cmd`, wondering what happens when the underlying implementation changes?

File: internal/job_directory.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+// StopJob stops the cmd
+func (d *JobDirectory) StopJob(uid model.UserID, jid model.JID) error {
+	cmd, err := d.retrieveJob(uid, jid)
+	if err != nil {
+		return err
+	}
+	// .Stop() safe to call concurrently and multiple times
+	// Note that only a started command can stopped, so there is a small time
+	// after adding a job where, in theory, stopping it would fail.
+	return cmd.Stop()
```
Comment (r0mant) [inline internal/job_directory.go:86]: What does `cmd.Stop()` do?

File: internal/server.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+		jobRequest := model.JobRequest{}
+		err = json.Unmarshal(bs, &jobRequest)
+		if err != nil {
+			log.Println("Error while parsing jobRequest.")
+			sendError(w, err)
+			return
+		}
+
+		command := cmd.NewCmd(jobRequest.Command, jobRequest.Arguments...)
```
Comment (r0mant) [inline internal/server.go:113]: I would consider moving the business logic of creating a command and adding to the registry to the job directory service.

File: internal/server.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+		// TODO: we could block until the process is actually started. We could
+		// then guarantee that a job said to be "running" to the client is either
+		// still running or done when the client receives the answer (edge case
+		// though).
+		// We leave like this for now.
+		command.Start()
```
Comment (r0mant) [inline internal/server.go:120]: Can this return an error?

File: internal/server.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+		if err != nil {
+			sendError(w, err)
+			return
+		}
+
+		err = jobDirectory.StopJob(uid, jid)
+		if err != nil {
+			sendError(w, err)
+			return
+		}
```
Comment (r0mant) [inline internal/server.go:142]: Does this send any reply to the client?

File: pkg/model.go [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+type JobLog struct {
+	Stdout string `json:"stdout"`
+	Stderr string `json:"stderr"`
+}
+
+// JID is the job ID type.
+type JID = uuid.UUID
+
+// UserID is the user ID type.
+type UserID = int
```
Comment (r0mant) [inline pkg/model.go:17]: Why do you need these type aliases, why not use types directly?

File: pkg/model.go [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+	ID        JID      `json:"id"`
+	Command   string   `json:"command"`
+	Arguments []string `json:"arguments"`
+	Status    Status   `json:"status"`
+	// Can be -1 in the stopped case.
+	ExitCode *int `json:"exitCode,omitempty"`
+}
+
+// JobRequest has the fields necessary to start a job. Arguments can be empty.
+type JobRequest struct {
```
Comment (r0mant) [inline pkg/model.go:40]: In Go it is not very customary to have dedicated "model" packages with all defined types, I would move them closer to where they are used.

File: internal/server.go (hunk 15 above)
Comment (hiqua) [reply internal/server.go:142]: It sends the error verbatim as the value of the field `err` in a json object. That means that it leaks information, but given the trust model I think it's fine, let me know otherwise.

File: pkg/model.go (hunk 16 above)
Comment (hiqua) [reply pkg/model.go:17]: I meant it as a way to check that JIDs are not used as UIDs, since they're actually both UUID (in the last PR where I adapt these types).

But I'm realizing that I used type aliases rather than type definitions, and I should have used the latter instead (so I should have done `type UserID int` instead of `type UserID = int`). I'll switch to this in this PR.

Let me know if you prefer to use the types directly. I think that's a common practice in most languages, right now I don't see why it would be different in Go, but you know better.

File: internal/server.go (hunk 14 above)
Comment (hiqua) [reply internal/server.go:120]: Not directly, but there is a possible error check to do in command.Status() which returns this struct:
```go
type Status struct {
	Cmd      string
	PID      int
	Complete bool     // false if stopped or signaled
	Exit     int      // exit code of process
	Error    error    // Go error
	StartTs  int64    // Unix ts (nanoseconds), zero if Cmd not started
	StopTs   int64    // Unix ts (nanoseconds), zero if Cmd not started or running
	Runtime  float64  // seconds, zero if Cmd not started
	Stdout   []string // buffered STDOUT; see Cmd.Status for more info
	Stderr   []string // buffered STDERR; see Cmd.Status for more info
}
```

I should at least do it when reporting on the job status, I think I forgot about this.

I could also check for an error already, but then it will be the same question as hinted at in the comment: unless I wait a bit, it could be empty just because the status was checked before the command was started. This would make the output inconsistent across runs of the same job.

I think it's fine to leave it to the caller to get the job status again once it has started the job, but let me know if you think otherwise.

File: internal/job_directory.go (hunk 12 above)
Comment (hiqua) [reply internal/job_directory.go:86]:
```go
// Stop stops the command by sending its process group a SIGTERM signal.
// Stop is idempotent. Stopping and already stopped command returns nil.
//
// Stop returns ErrNotStarted if called before Start or StartWithStdin. If the
// command is very slow to start, Stop can return ErrNotStarted after calling
// Start or StartWithStdin because this package is still waiting for the system
// to start the process. All other return errors are from the low-level system
// function for process termination.
```

File: pkg/model.go (hunk 17 above)
Comment (hiqua) [reply pkg/model.go:40]: I separated it from `internal/` so that they can be reused, e.g. if someone uses the REST API, they should be able to use these types more easily. But that's probably overkill here, I'll merge the types with the rest as you suggest.

File: cmd/server/server.go [hunk 18]
```diff
@@ -0,0 +1,11 @@
+package main
+
+import (
+	"log"
+
+	"github.com/hiqua/rworker/internal"
+)
+
+func main() {
+	log.Fatal(server.Serve())
```
[the remark this answers is not in the document]
Comment (hiqua) [reply cmd/server/server.go:10]: Filed as #6 and I'll address this in the last PR, if I move files around now the rebasing might be needlessly complicated and the diff harder to parse.

File: pkg/model.go (hunk 17 above)
Comment (hiqua) [reply pkg/model.go:40]: Filed as #7, will do that in the last PR.

File: internal/job_directory.go (hunk 10 above)
Comment (hiqua) [reply internal/job_directory.go:86]: I assumed that trailing `\n` would all be removed, but actually my original comment was wrong, and none of it is removed. I'm removing my comment.

File: internal/job_directory.go (hunk 11 above)
Comment (hiqua) [reply internal/job_directory.go:60]: By the way this code was missing a check on Status.Error. I've added it in a later change.

File: internal/job_directory.go [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+// JobDirectory provides a safe interface to the map containing all jobs for all users.
+type JobDirectory struct {
+	jobs map[model.UserID]map[model.JID]*cmd.Cmd
+	// One global lock. Other designs could perform.
+	lock sync.RWMutex
+}
+
+// NewJobDirectory initializes a new Job Directory.
+func NewJobDirectory() JobDirectory {
+	return JobDirectory{jobs: make(map[model.UserID]map[model.JID]*cmd.Cmd), lock: sync.RWMutex{}}
```
Comment (webvictim) [inline internal/job_directory.go:21]: This is quite hard to read. As @r0mant says above, it would be nicer to use a struct to represent the job and pass that around.

File: internal/job_directory_test.go [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"testing"
+
+	model "github.com/hiqua/rworker/pkg"
```
Comment (webvictim) [inline internal/job_directory_test.go:11]: Calling this import `model` when the actual package imported is called `pkg` is a little confusing. It can work in a smaller project like this but would quickly get hard to follow with a bigger one. I would look at making this consistent across the two - either rename `pkg` to `model` and import that, or import `pkg` instead.

File: internal/job_directory_test.go [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func generateTestingFolder(n int) (string, error) {
+	// We don't delete the temporary folder.
+	dir, err := ioutil.TempDir("", "prefix")
+	if err != nil {
+		return "", err
+	}
+	for i := 0; i < n; i++ {
+		err = os.Mkdir(dir+"/folder"+fmt.Sprint(i), 0755)
```
Comment (webvictim) [inline internal/job_directory_test.go:83]: Using [filepath.Join()](https://golang.org/pkg/path/filepath/#Join) would be better for this.

File: internal/server.go [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func addJob(jobDirectory *JobDirectory) func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+		uid := dummyExtractUserID()
+		logOperation("addJob", uid)
+
+		bs, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Println("Error while reading body.")
```
[the remark this answers is not in the document]
Comment (hiqua) [reply internal/server.go:96]: Error sent to the user as json output.

File: internal/job_directory.go (hunk 11 above)
Comment (hiqua) [reply internal/job_directory.go:60]: Yes the relevant package for this would be: https://golang.org/src/syscall/syscall_linux.go?s=8206:8242#L326

I'm not sure there are guarantees coming with this package. On the other hand, breaking changes in this package for exposed function would break a lot of things. I think in this case, for a real project, I would ask clarification and guarantees from the library (cmd) maintainer, and possibly add a test myself both to the library and this project that this `-1` is indeed returned in case of signal.
